### PR TITLE
Feature/requested lesson sorting algorithm

### LINKED ISSUE #132: Implement a filtering algorithm for Requested lesson, to ease lesson finding for Tutors

### PR DESCRIPTION
**Pull request by alexs-rufer**
## 🎯 PR Description

This pull request introduces the suitability score computation (the "lesson sorting algorithm") and its integration into the `RequestedLessonsScreen`. The suitability score evaluates how well a tutor matches a lesson request based on schedule compatibility, academic level, price range, and location proximity. Here is a description of how the score is computed, we can still discuss about the weights and different hyper parameters values!
⚠️ Do not confuse this algorithm with the "Matching algorithm" that matches a lesson to a tutor! In this PR, its just sorting the lessons to ease lesson picking from the tutor side :)

## 🧮 Score Computation

The suitability score is computed as a weighted combination of 4 parameters:

**1. Schedule Compatibility (25%)**:  
  Evaluates if the tutor is available during the scheduled lesson time. Normalized score: 0 if unavailable, 1 if fully available.

**2. Academic Level Compatibility (35%)**:  
   Assesses the tutor's academic level compared to the student's. Tutors with higher levels score better, normalized between 0 (incompatible) and 1 (maximum compatibility). If the tutor is the same academic level as the student, the score is 0.3 and 0 if under.

**3. Price Compatibility (25%)**:  
   Checks if the tutor's price falls within the lesson's min and max price range. Scores linearly decrease for prices outside the range.

**4. Location Proximity (15%)**:  
   This is not implemented yet, as it seems a bit more complicated and I should dedicate a whole PR for this one. I added a placeholder function. It will compute the physical proximity between the tutor and the lesson location. Current implementation returns a fixed score of 1.

The overall score is computed as:  
`Score = (0.25 * Schedule) + (0.35 * Academic Level) + (0.25 * Price) + (0.15 * Location)`  

All scores are scaled to a percentage (0–100%).

## 📂 File Added

 **SuitabilityScoreCalculator.kt**
- I created a new `SuitabilityScoreCalculator` object to compute suitability scores, as described above. 
- Added a `getColorForScore` function to generate smooth gradient colors based on scores, with adjustments for light and dark themes (used in `DisplayLessons.kt`).
- Added helper functions `computeLanguageMatch` and `computeSubjectMatch` for filtering (used in `RequestedLessons.kt`).

## 📂 Files Changed

 **Lesson.kt**
- Added a `parseLessonDate` method to convert a lesson's `timeSlot` into a `LocalDateTime`.

 **DisplayLessons.kt**
- Added the `suitabilityScore` parameter to display the computed score on lesson cards.
- Updated the lesson card UI to show "Recommended at X%" in bold, with smooth colouring gradient between green and red.

 **RequestedLessons.kt**
- Modified filtering logic to:
  - Exclude lessons where the tutor's subject or language does not match (this was not done before, and we discussed about it with @AlexisAllmnd ).
  - Sort lessons by suitability score (descending).

## 🧪 How to Test
You can log in as a tutor, click on "Find student" and play around with the parameters of your account to see different scores (you can change your academic level, schedule, the languages you speak, the subjects you teach etc..). Do not hesitate to tell me if you feel like some lessons are recommended to you but should not be (and vice versa), as I said before we can play with the weights!

## 📸 Screenshots
You can see here the different scores computed when I am looking for a lesson as a tutor, in light and dark theme.

image

image

Close #132

### CODE REVIEW

File: app/src/main/java/com/github/se/project/ui/lesson/RequestedLessons.kt [hunk 1]
```diff
@@ -66,9 +72,30 @@ fun RequestedLessonsScreen(
...
+          .mapNotNull { lesson ->
+            val studentProfile =
+                listProfilesViewModel.profiles.value.find { it.uid == lesson.studentUid }
+                    ?: return@mapNotNull null
+            val tutorProfile = currentProfile ?: return@mapNotNull null
+
+            val score =
+                SuitabilityScoreCalculator.calculateSuitabilityScore(
+                    lesson, tutorProfile, studentProfile)
+            lesson to score
```
Comment (MPepito) [inline app/src/main/java/com/github/se/project/ui/lesson/RequestedLessons.kt:96]: Great declarative implementation of the filtering, easy to read and understand

**Review by MPepito — APPROVED** [review]
This feature was 100% your idea and you converted this idea into this useful and well designed feature. The code is easy to understand and well modularized. Well done! I added a few suggestions for some optimisations, up to you to implement it now, I am ready to re approve once you commit any adjustments

File: app/src/main/java/com/github/se/project/model/lesson/Lesson.kt [hunk 2]
```diff
@@ -18,4 +20,18 @@ data class Lesson(
...
+) {
+  /**
+   * Parses the lesson's timeSlot into a LocalDateTime object.
+   *
+   * @return The LocalDateTime representation of the timeSlot, or null if parsing fails.
+   */
+  fun parseLessonDate(): LocalDateTime? {
+    return try {
+      val formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy'T'HH:mm:ss")
+      LocalDateTime.parse(timeSlot, formatter)
```
[the remark this answers is not in the document]
Comment (alexs-rufer) [reply app/src/main/java/com/github/se/project/model/lesson/Lesson.kt:32]: I totally agree with you, I would have implemented it that way, but I just used what was already implemented :(

File: app/src/main/java/com/github/se/project/model/lesson/SuitabilityScoreCalculator.kt [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+    val levelDifference = tutorLevelIndex - studentLevelIndex
+
+    if (levelDifference < 0) return 0.0
+
+    val maxLevelDiff = AcademicLevel.entries.size - studentLevelIndex
+    // Ensure the level difference is within bounds [0, maxLevelDiff]
+    val normalizedDifference =
+        levelDifference.coerceIn(0, maxLevelDiff).toDouble() / maxLevelDiff.toDouble()
+
+    val score = 0.3 + 0.7 * normalizedDifference
```
[the remark this answers is not in the document]
Comment (alexs-rufer) [reply app/src/main/java/com/github/se/project/model/lesson/SuitabilityScoreCalculator.kt:77]: nothing, as I explained in my PR it was just that it seemed fair that a score is 0.3 if same academic level and 0 if under

File: app/src/main/java/com/github/se/project/model/lesson/SuitabilityScoreCalculator.kt [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    }
+
+    val tutorIsAvailable = tutorProfile.schedule[dayOfWeekIndex - 1][hourIndex] == 1
+
+    return if (tutorIsAvailable) 1.0 else 0.0
+  }
+
+  // TODO: this function might be computationally costly to retrieve every student profile.
+  // I would like to ask the coaches
+  private fun computeAcademicLevelCompatibility(
```
[the remark this answers is not in the document]
Comment (alexs-rufer) [reply app/src/main/java/com/github/se/project/model/lesson/SuitabilityScoreCalculator.kt:58]: I agree with you, we discussed about that with @AlexisAllmnd and again I feel like we should determine this with the group if that's ok for you! :)

**Review by AlexisAllmnd — COMMENTED** [review]
Approvement after CI passed

**Review by AlexisAllmnd — APPROVED** [review]

**Pull request merged**